### PR TITLE
Support image sizes in news and calendar feeds

### PR DESCRIPTION
**Pull request by bezin**
Implements #1842

### CODE REVIEW

File: calendar-bundle/src/Resources/contao/dca/tl_calendar_feed.php [hunk 1]
```diff
@@ -211,7 +211,19 @@
 			'inputType'               => 'textarea',
 			'eval'                    => array('style'=>'height:60px', 'tl_class'=>'clr'),
 			'sql'                     => "text NULL"
-		)
+		),
+		'imageSize' => array
```
Comment (fritzmg) [inline calendar-bundle/src/Resources/contao/dca/tl_calendar_feed.php]: In `tl_module`, the field is called `imgSize` and everywhere else just `size`. I think we should use either of these (may be `imgSize`) in order to have more consistency.

**Review by fritzmg — CHANGES_REQUESTED** [review]

**Review by fritzmg — DISMISSED** [review]
LGTM 👍

File: calendar-bundle/src/Resources/contao/languages/en/tl_calendar_feed.xlf [hunk 2]
```diff
@@ -56,6 +56,12 @@
       <trans-unit id="tl_calendar_feed.description.1">
         <source>Here you can enter a short description of the calendar feed.</source>
       </trans-unit>
+      <trans-unit id="tl_calendar_feed.imageSize.0">
+        <source>Image size</source>
+      </trans-unit>
+      <trans-unit id="tl_calendar_feed.imageSize.1">
```
Comment (aschempp) [inline calendar-bundle/src/Resources/contao/languages/en/tl_calendar_feed.xlf]:
```suggestion
      <trans-unit id="tl_calendar_feed.imgSize.0">
        <source>Image size</source>
      </trans-unit>
      <trans-unit id="tl_calendar_feed.imgSize.1">
```

File: core-bundle/src/Resources/contao/library/Contao/FeedItem.php [hunk 3]
```diff
@@ -113,12 +117,29 @@ public function addEnclosure($strFile, $strUrl=null, $strMedia='enclosure')
...
+		{
+			$image = System::getContainer()->get('contao.image.image_factory')->create($rootDir . '/' . $strFile, $imageSize);
+
+			if ($image instanceof DeferredImageInterface)
+			{
+				System::getContainer()->get('contao.image.resizer')->resizeDeferredImage($image);
+			}
+
+			$fileUrl = $strUrl . System::urlEncode($image->getUrl($rootDir));
+			$relativeFilePath = str_replace($rootDir . '/', '', $image->getPath());
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/library/Contao/FeedItem.php]: this should probably use `PathUtils::makeRelative()` (not sure if that's the exact name 🙈 )

File: news-bundle/src/Resources/contao/languages/en/tl_news_feed.xlf [hunk 4]
```diff
@@ -56,6 +56,12 @@
       <trans-unit id="tl_news_feed.description.1">
         <source>Here you can enter a short description of the news feed.</source>
       </trans-unit>
+      <trans-unit id="tl_news_feed.imageSize.0">
+        <source>Image size</source>
+      </trans-unit>
+      <trans-unit id="tl_news_feed.imageSize.1">
```
Comment (aschempp) [inline news-bundle/src/Resources/contao/languages/en/tl_news_feed.xlf]:
```suggestion
      <trans-unit id="tl_news_feed.imgSize.0">
        <source>Image size</source>
      </trans-unit>
      <trans-unit id="tl_news_feed.imgSize.1">
```

**Review by aschempp — CHANGES_REQUESTED** [review]

File: core-bundle/src/Resources/contao/library/Contao/FeedItem.php (hunk 3 above)
Comment (m-vo) [reply core-bundle/src/Resources/contao/library/Contao/FeedItem.php]: Yup, it's called `Path`: :slightly_smiling_face: 

```php
$relativeFilePath = Path::makeRelative($image->getPath(), $rootDir);
```

File: core-bundle/src/Resources/contao/library/Contao/FeedItem.php [hunk 5]
```diff
@@ -113,12 +118,29 @@ public function addEnclosure($strFile, $strUrl=null, $strMedia='enclosure')
 			$strUrl = Environment::get('base');
 		}
 
+		$fileUrl = $strUrl . System::urlEncode($strFile);
 		$objFile = new File($strFile);
 
+		$imageSize = StringUtil::deserialize($imageSize, true);
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/library/Contao/FeedItem.php]: This line is overwriting the method argument, which is not a good thing to do. Maybe you can use a different name for the unserialized content?

File: calendar-bundle/src/Resources/contao/dca/tl_calendar_feed.php [hunk 6]
```diff
@@ -211,7 +211,19 @@
...
+		'imgSize' => array
+		(
+			'exclude'                 => true,
+			'inputType'               => 'imageSize',
+			'reference'               => &$GLOBALS['TL_LANG']['MSC'],
+			'eval'                    => array('rgxp'=>'natural', 'includeBlankOption'=>true, 'nospace'=>true, 'helpwizard'=>true, 'tl_class'=>'w50'),
+			'options_callback' => static function ()
+			{
+				return System::getContainer()->get('contao.image.image_sizes')->getOptionsForUser(BackendUser::getInstance());
+			},
```
Comment (m-vo) [inline calendar-bundle/src/Resources/contao/dca/tl_calendar_feed.php]: What about putting this in a class, tagging it with…

```
@Callback(table="tl_calendar_feed ", target="fields.imgSize.options")
@Callback(table="tl_news_feed ", target="fields.imgSize.options")
```
… and using proper DI for both instances?

Comment (bezin) [reply calendar-bundle/src/Resources/contao/dca/tl_calendar_feed.php]: I really dislike the tagging pattern for select fields: It completely removes any reference to the options from the field definition: You look at the configuration and have no clue a) which options it has or b) even where these are compiled. You have to find the implementation manually somewhere. I would prefer defining a callabe over tagging if we want to remove the anonymous function. 

However, if this it where Contao wants to go I will bow to that.

Comment (SeverinGloeckle) [reply calendar-bundle/src/Resources/contao/dca/tl_calendar_feed.php]: You are definitely right that the loss of reference between an option field and its options is inconvenient.

The notation of a callable / service call can at least provide a hint on where to look for the options source (e.g. `contao.image.image_sizes`), but that still requires a manual lookup of the service ID and a public service - so it's not perfect either.

For the specific use-case of image sizes, there probably should be an easier way to get the options for the current backend user. Either the `ImageSizes` service should know about the current request and optionally infer the current backend user from it or a more general `BackendUserSettings` service should be added which can be queried for options and settings of the current backend user.

Alternatively, since the `ImageSize` widget is already querying the available options from the image sizes service in its validator: Maybe this logic can also be used as a default so you don't have to define the options for "regular" image size options at all?

However, as the current one-line callback to get the image sizes is used in multiple places already, I think it should be used here, too (for now, and maybe get refactored later together with similar callbacks).

Comment (SeverinGloeckle) [reply calendar-bundle/src/Resources/contao/dca/tl_calendar_feed.php]: As a side note: In my opinion, the specification of a closure as a callback should be completely discouraged in the long term, as these can only be defined within PHP code. A conversion / dumping of a DCA into another format - for example YAML - as a result would be impossible.

So that's definitely something to think about in the future.

Comment (bezin) [reply calendar-bundle/src/Resources/contao/dca/tl_calendar_feed.php]: >  but that still requires a manual lookup of the service ID and a public service - so it's not perfect either.

Yes, true – not perfect, but at least I have a service ID to search for in my IDE.

> However, as the current one-line callback to get the image sizes is used in multiple places already, I think it should be used here, too (for now, and maybe get refactored later together with similar callbacks).

That is what I think, too.

File: core-bundle/src/Resources/contao/library/Contao/FeedItem.php (hunk 5 above)
Comment (leofeyer) [reply core-bundle/src/Resources/contao/library/Contao/FeedItem.php]: aa374353b51684cecff5afe6fff22836b5e07b5f did not fix this issue, I'm afraid, because now both the argument and the variable are called `$size`. One should be called `$imageSize` and the other one `$size`.

File: core-bundle/src/Resources/contao/library/Contao/FeedItem.php [hunk 7]
```diff
@@ -101,9 +104,11 @@ public function __isset($strKey)
 	 * @param string $strUrl   The base URL
 	 * @param string $strMedia The media type
 	 */
-	public function addEnclosure($strFile, $strUrl=null, $strMedia='enclosure')
+	public function addEnclosure($strFile, $strUrl=null, $strMedia='enclosure', $size = null)
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/library/Contao/FeedItem.php]: `$size` is missing in the docbloc

File: core-bundle/src/Resources/contao/library/Contao/FeedItem.php [hunk 8]
```diff
@@ -113,12 +118,29 @@ public function addEnclosure($strFile, $strUrl=null, $strMedia='enclosure')
...
 
+		$size = StringUtil::deserialize($size, true);
+
+		if ($size && $objFile->isImage)
+		{
+			$image = System::getContainer()->get('contao.image.image_factory')->create($rootDir . '/' . $strFile, $size);
+
+			if ($image instanceof DeferredImageInterface)
+			{
+				System::getContainer()->get('contao.image.resizer')->resizeDeferredImage($image);
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/library/Contao/FeedItem.php]: why do we need to force-resize deferred images? They are not embedded in the feed so they would just be created on request?

**Review by aschempp — CHANGES_REQUESTED** [review]

File: core-bundle/src/Resources/contao/library/Contao/FeedItem.php (hunk 8 above)
Comment (bezin) [reply core-bundle/src/Resources/contao/library/Contao/FeedItem.php]: True, this is probably specific to my original use case. We used the RSS feed to import news articles into and send them with CleverReach. Their client did not wait long enough for the images to be created and just did not load them at all. 

We could make DeferredImages the default, but I would like to add a checkbox then, that allows me to create images on feed creation. WDYT?

Comment (leofeyer) [reply core-bundle/src/Resources/contao/library/Contao/FeedItem.php]: @ausi /cc

Comment (aschempp) [reply core-bundle/src/Resources/contao/library/Contao/FeedItem.php]: if you have that specific problem, you should configure that in your application – disable deferred images or run the image-generation command in a cron job.

Comment (ausi) [reply core-bundle/src/Resources/contao/library/Contao/FeedItem.php]: > but I would like to add a checkbox then, that allows me to create images on feed creation. WDYT?

I would not add this checkbox as this requirement is a quite specific use-case IMO.

But we should make sure that your use-case is possible at least. Is there an event or hook that you can use to intercept the feed contents? If so, it might be possible to resize the images like the newsletter module does with `$objFile->exists() || $objFile->createIfDeferred()`: https://github.com/contao/contao/blob/4299a5116b73cdbcd024c2d86b0533ae15ec3137/core-bundle/src/Resources/contao/library/Contao/Email.php#L476

Comment (bezin) [reply core-bundle/src/Resources/contao/library/Contao/FeedItem.php]: Agreed, I will look into that. As far as I see, there is no Hook/Event yet, so we would have to add one. I will figure out, where would be the best place.

File: core-bundle/src/Resources/contao/library/Contao/FeedItem.php (hunk 7 above)
Comment (bezin) [reply core-bundle/src/Resources/contao/library/Contao/FeedItem.php]: Fixed

File: core-bundle/src/Resources/contao/library/Contao/FeedItem.php [hunk 9]
```diff
@@ -113,12 +118,23 @@ public function addEnclosure($strFile, $strUrl=null, $strMedia='enclosure')
...
 		$objFile = new File($strFile);
 
+		$size = StringUtil::deserialize($imageSize, true);
+
+		if ($size && $objFile->isImage)
+		{
+			$image = System::getContainer()->get('contao.image.image_factory')->create($rootDir . '/' . $strFile, $size);
+			$fileUrl = $strUrl . System::urlEncode($image->getUrl($rootDir));
+			$relativeFilePath = Path::makeRelative($image->getPath(), $rootDir);
+			$objFile = new File($relativeFilePath);
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/library/Contao/FeedItem.php]: Are you sure to use `File` class for a non-resized image?

File: core-bundle/src/Resources/contao/library/Contao/FeedItem.php (hunk 8 above)
Comment (m-vo) [reply core-bundle/src/Resources/contao/library/Contao/FeedItem.php]: Alternatively: If we migrate the feed composition to templates, this would be easily adjustable for users as well.

File: core-bundle/src/Resources/contao/library/Contao/FeedItem.php (hunk 9 above)
Comment (ausi) [reply core-bundle/src/Resources/contao/library/Contao/FeedItem.php]: I don’t understand the question. It overwrites the `$objFile` variable from above which seems to be correct.

File: core-bundle/src/Resources/contao/library/Contao/FeedItem.php (hunk 8 above)
Comment (aschempp) [reply core-bundle/src/Resources/contao/library/Contao/FeedItem.php]: > Alternatively: If we migrate the feed composition to templates, this would be easily adjustable for users as well.

not related to this PR, and something we might look at when rewriting to a controller 😉

File: core-bundle/src/Resources/contao/library/Contao/FeedItem.php (hunk 9 above)
Comment (bezin) [reply core-bundle/src/Resources/contao/library/Contao/FeedItem.php]: As far as I see the file is not required to be present, when instantiating `File`? However, I added a check if the file exists, before I get `$objFile->size` some lines below to avoid a warning

File: calendar-bundle/src/Resources/contao/dca/tl_calendar_feed.php (hunk 6 above)
Comment (leofeyer) [reply calendar-bundle/src/Resources/contao/dca/tl_calendar_feed.php]: I have created a separate issue for discussing this: #3907

File: core-bundle/src/Resources/contao/library/Contao/FeedItem.php (hunk 8 above)
Comment (leofeyer) [reply core-bundle/src/Resources/contao/library/Contao/FeedItem.php]: Resolved through 1883a64eb8fcdaa29eef11f9c6eb295a6c096fd8

File: calendar-bundle/src/Resources/contao/dca/tl_calendar_feed.php (hunk 6 above)
Comment (bezin) [reply calendar-bundle/src/Resources/contao/dca/tl_calendar_feed.php]: I am currently refactoring this: I implement a general purpose listener that is used in all other places in the core, too (`tl_layout`, `tl_module`, `tl_content`). I then use this in `tl_news` and `tl_news_feed` as well as `tl_calendar_event`  and `tl_calendar_feed`.

**Review by leofeyer — DISMISSED** [review]

File: core-bundle/src/Resources/contao/library/Contao/FeedItem.php [hunk 10]
```diff
@@ -113,15 +118,29 @@ public function addEnclosure($strFile, $strUrl=null, $strMedia='enclosure')
 			$strUrl = Environment::get('base');
 		}
 
+		$fileUrl = $strUrl . System::urlEncode($strFile);
 		$objFile = new File($strFile);
+		$size = StringUtil::deserialize($imageSize, true);
+
+		if ($size && $objFile->isImage)
+		{
+			$image = System::getContainer()->get('contao.image.image_factory')->create($rootDir . '/' . $strFile, $size);
```
Comment (m-vo) [inline core-bundle/src/Resources/contao/library/Contao/FeedItem.php]:
```suggestion
			$image = System::getContainer()->get('contao.image.image_factory')->create(Path::join($rootDir, $strFile), $size);
```

File: core-bundle/src/Resources/contao/library/Contao/FeedItem.php [hunk 11]
```diff
@@ -97,13 +99,16 @@ public function __isset($strKey)
...
+	 * @param string $strMedia  The media type
+	 * @param mixed  $imageSize The image size
 	 */
-	public function addEnclosure($strFile, $strUrl=null, $strMedia='enclosure')
+	public function addEnclosure($strFile, $strUrl=null, $strMedia='enclosure', $imageSize = null)
 	{
-		if (!$strFile || !file_exists(System::getContainer()->getParameter('kernel.project_dir') . '/' . $strFile))
+		$rootDir = System::getContainer()->getParameter('kernel.project_dir');
+
+		if (!$strFile || !file_exists($rootDir . '/' . $strFile))
```
Comment (m-vo) [inline core-bundle/src/Resources/contao/library/Contao/FeedItem.php]:
```suggestion
		if (!$strFile || !file_exists(Path::join($rootDir, $strFile)))
```

File: core-bundle/src/Resources/contao/library/Contao/FeedItem.php (hunk 7 above)
Comment (bezin) [reply core-bundle/src/Resources/contao/library/Contao/FeedItem.php]: https://github.com/contao/contao/pull/3630/commits/18bac58c6522ab73dcea26867d8e851f1cdb1bda

**Review by m-vo — DISMISSED** [review]

File: core-bundle/src/Resources/contao/dca/tl_content.php [hunk 12]
```diff
@@ -265,10 +265,6 @@
 			'inputType'               => 'imageSize',
 			'reference'               => &$GLOBALS['TL_LANG']['MSC'],
 			'eval'                    => array('rgxp'=>'natural', 'includeBlankOption'=>true, 'nospace'=>true, 'helpwizard'=>true, 'tl_class'=>'w50'),
-			'options_callback' => static function ()
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/dca/tl_content.php:268]: Why did you remove the options callback here (and in tl_layout and tl_module)?

Comment (bezin) [reply core-bundle/src/Resources/contao/dca/tl_content.php:268]: Because I added a [general purpose listener for this and use it via annotations](https://github.com/contao/contao/pull/3630/files#diff-d71a0a24c80180705e7fd49548488f67bc086cb5cc0e547dea38794fb68428f5R21-R23)

**Review by aschempp — DISMISSED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @bezin.
